qrbill/bill.py

```diff
@@ -498,45 +540,52 @@ def add_header(text):
                 extra_infos = [self.extra_infos]
             # TODO: handle line breaks for long infos (mandatory 5mm margin)
             for info in wrap_infos(extra_infos):
-                dwg.add(dwg.text(info, (payment_detail_left, '%smm' % y_pos), **self.font_info))
+                grp.add(dwg.text(info, (payment_detail_left, mm(y_pos)), **self.font_info))
                 y_pos += line_space
 
         if self.debtor:
             add_header(self.label("Payable by"))
             for line_text in self.debtor.as_paragraph():
-                dwg.add(dwg.text(line_text, (payment_detail_left, '%smm' % y_pos), **self.font_info))
+                grp.add(dwg.text(line_text, (payment_detail_left, mm(y_pos)), **self.font_info))
                 y_pos += line_space
         else:
             add_header(self.label("Payable by (name/address)"))
             # The specs recomment at least 2.5 x 6.5 cm
             self.draw_blank_rect(
-                dwg, x=payment_detail_left, y='%smm' % y_pos,
-                width='65mm', height='25mm'
+                dwg, grp, x=payment_detail_left, y=mm(y_pos),
+                width=mm(65), height=mm(25)
             )
             y_pos += 28
 
         if self.final_creditor:
             add_header(self.label("In favor of"))
             for line_text in self.final_creditor.as_paragraph():
-                dwg.add(dwg.text(line_text, (payment_detail_left, '%smm' % y_pos), **self.font_info))
+                grp.add(dwg.text(line_text, (payment_detail_left, mm(y_pos)), **self.font_info))
                 y_pos += line_space
 
         if self.due_date:
             add_header(self.label("Payable by "))
-            dwg.add(dwg.text(
-                format_date(self.due_date), (payment_detail_left, '%smm' % y_pos), **self.font_info
+            grp.add(dwg.text(
+                format_date(self.due_date), (payment_detail_left, mm(y_pos)), **self.font_info
             ))
             y_pos += line_space
-
-        if isinstance(file_out, (str, Path)):
-            dwg.saveas(file_out)
-        else:
-            dwg.write(file_out)
+        return grp
 
 
 def add_mm(*mms):
     """Utility to allow additions of '23mm'-type strings."""
-    return '%smm' % str(sum(float(mm[:-2]) for mm in mms))
+    return sum(
+        mm(float(m[:-2])) if isinstance(m, str) else m for m in mms
+    )
+
+
+def mm(val):
+    """Convert val (as mm, either number of '12mm' str) into user units."""
+    try:
+        val = float(val.rstrip('mm'))
+    except AttributeError:
+        pass
+    return round(val * MM_TO_UU, 2)
```

Comment:
I don't think it matters, but shouldn't we try to keep the precision instead of rounding?

Reply:
Maybe 5 just for good measure :)